src/core/lib/transport/handshaker.cc

```diff
@@ -201,10 +197,15 @@ void HandshakeManager::DoHandshake(grpc_endpoint* endpoint,
     GRPC_CLOSURE_INIT(&on_handshake_done_, on_handshake_done, &args_,
                       grpc_schedule_on_exec_ctx);
     // Start deadline timer, which owns a ref.
-    Ref().release();
-    GRPC_CLOSURE_INIT(&on_timeout_, &HandshakeManager::OnTimeoutFn, this,
-                      grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&deadline_timer_, deadline, &on_timeout_);
+    const Duration time_to_deadline = deadline - Timestamp::Now();
+    deadline_timer_handle_ =
+        event_engine_->RunAfter(time_to_deadline, [self = Ref()]() mutable {
+          ApplicationCallbackExecCtx callback_exec_ctx;
+          ExecCtx exec_ctx;
+          self->Shutdown(GRPC_ERROR_CREATE("Handshake timed out"));
+          // HandshakeManager deletion might require an active ExecCtx.
```

Comment:
Interesting! Are there tests that fail if `self` isn't released before the end of the callback?

Reply:
This is mostly based on the code here: https://github.com/grpc/grpc/blob/05ab5c08a468a1555b07fd9a70cb9edc5bb849d7/src/core/ext/filters/client_channel/subchannel.cc#L928

The closure holds a ref to the HandshakeManager and when it is destructed it triggers a series of destruction steps on the event_engine thread which may need an ExecCtx to be live on the thread. Thus, self.reset() triggers this process before the previous ExecCtx is destructed. Otherwise we may see this crash in one of the tests: https://paste.googleplex.com/4627665280565248

Reply:
Wrong paste link sorry... This should be the one: https://paste.googleplex.com/5265162796793856

Reply:
Got it, thank you!